packages/site/cms-site/src/iframebridge/IFrameBridge.tsx

```diff
@@ -30,13 +54,82 @@ export const IFrameBridgeContext = React.createContext<IFrameBridgeContext>({
     const [block, setBlock] = React.useState<unknown | undefined>(undefined);
     const [selectedAdminRoute, setSelectedAdminRoute] = React.useState<string | undefined>(undefined);
     const [hoveredAdminRoute, setHoveredAdminRoute] = React.useState<string | undefined>(undefined);
     const [showOutlines, setShowOutlines] = React.useState<boolean>(false);
+    const [previewElements, setPreviewElements] = React.useState<PreviewElement[]>([]);
+    const [recalculatePreviewDataIndex, setRecalculatePreviewDataIndex] = React.useState<number>(0);
+
+    const childrenWrapperRef = React.useRef<HTMLDivElement>(null);
+
+    const triggerRecalculationOfPreviewData = React.useCallback(() => {
+        setRecalculatePreviewDataIndex((index) => index + 1);
+    }, []);
+
+    React.useEffect(() => {
+        const mutationObserver = new MutationObserver(() => {
+            triggerRecalculationOfPreviewData();
+        });
+
+        const resizeObserver = new ResizeObserver(() => {
+            triggerRecalculationOfPreviewData();
+        });
+
+        if (childrenWrapperRef.current) {
+            mutationObserver.observe(childrenWrapperRef.current, { childList: true, subtree: true });
+            resizeObserver.observe(childrenWrapperRef.current);
+        }
+
+        return () => {
+            mutationObserver.disconnect();
+            resizeObserver.disconnect();
+        };
+    }, [triggerRecalculationOfPreviewData]);
+
+    const calculatedPreviewElementsData: OverlayElementData[] = React.useMemo(() => {
+        return previewElements
+            .map((previewElement) => {
+                const childNodes = getRecursiveChildrenOfPreviewElement(previewElement.element);
+                const positioning = getCombinedPositioningOfElements(childNodes);
+
+                return {
+                    adminRoute: previewElement.adminRoute,
+                    label: previewElement.label,
+                    position: {
+                        top: positioning.top,
+                        left: positioning.left,
+                        width: positioning.width,
+                        height: positioning.height,
+                    },
+                };
+            })
+            .sort((previousElementData, nextElementData) => {
+                const previousSize = previousElementData.position.width * previousElementData.position.height;
+                const nextSize = nextElementData.position.width * nextElementData.position.height;
+                return nextSize - previousSize;
+            })
+            .map((elementData, index) => {
+                return {
+                    ...elementData,
+                    position: {
+                        ...elementData.position,
+                        zIndex: index + 1,
+                    },
+                };
+            });
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [previewElements, recalculatePreviewDataIndex]);
```

Comment:
Instead of using useMemo, a trigger state and disabling the ESLint rule, why not recalculate in the useEffect hook?

Reply:
Haha, no idea 😅 
I'm now calling the recalculation function directly when the observers trigger.